### ISSUE TITLE
Local machine learning of posts

### ISSUE THREAD

**Issue by amiyatulu** [mastodon/mastodon#22359]
### Pitch

The current feed causes information overload even if followers are just 50. 
If you follow an account that make posts every hour, user feed is filled with posts from that single account. 
It could be nice if we could have another feed that is trained by user building a personalized feed. Training can be done locally and algorithm and model is made open source/foss.
Yes it be implemented as a 3rd party app using the REST API.
I can't give exact steps as I have limited experience on machine learning.

### Motivation

It will help to reduce the information overload, and improve user experience.

**Comment by schmate** [comment]
@amiyatulu I do not know the Mastodon backend, I'm just a Data Scientist who loves the idea and the implementation of Mastodon.

Let's distinguish between two things: the algorithm and the model. The algorithm is the thing that learns, while the model is the set of rules it learns on historical data and applies on future data. (WARNING: Oversimplified but helpful explanation.)

Do you imagine this as an open source algorithm (packed with some other required tools) and a personal model for each user, trained locally, without the ranking model or any otherwise not stored user-content-interaction data being shared with the server?

**Comment by amiyatulu** [comment]
Yes. That's what I mean. But there are problems with training locally, because app may hang due to additional computation and model storage. Not sure how much computational resource it requires, as it depends on the algorithm.  
I think it will be not a problem to host the model on an API server if all code is open-source, and we can validate that the server or docker instance is running the same open-source code.

**Comment by pkreissel** [comment]
Don't know why this is getting downvotes. You could in theory already implement this with mastodons API. For example get the entire feed and timeline for a user and then rank according to past interactions etc. However I agree it would be good if there were some form of "plugin" space within mastodon itself to insert custom feed algorithms.

**Comment by amiyatulu** [comment]
It really annoying when you have a follower that posts every minute, and the feed is just filled with all their post. I think machine learning is even not required, just allowing to "decrease the intensity" of the posts of an account will do, where a few randomly selected recent posts are loaded to feed from the account.

**Comment by pkreissel** [comment]
I started to make a Mastodon-API powered Version of this. Feel free to contribute.
https://github.com/pkreissel/fedifeed

**Comment by pabloscloud** [comment]
> I started to make a Mastodon-API powered Version of this. Feel free to contribute.
> https://github.com/pkreissel/fedifeed

this but as view inside an app. Maybe if someone could provide this as api apps could implement it? @pkreissel

**Comment by amiyatulu** [comment]
@pkreissel Feeds are much better. Thanks. https://github.com/pkreissel/foryoufeed
